File: LeetCode/889. 根据前序和后序遍历构造二叉树.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def constructFromPrePost(self, preorder: List[int], postorder: List[int]) -> TreeNode:
        if len(preorder) == 0:
            return None
        l = len(preorder)
        val = preorder[0]
        if l == 1:
            return TreeNode(val)
        valOfLeftChild = preorder[1]
        postorderIndex = postorder.index(valOfLeftChild)
        leftCount = postorderIndex + 1
        left = self.constructFromPrePost(preorder[1:1 + leftCount], postorder[:leftCount])
        right = self.constructFromPrePost(preorder[1 + leftCount:], postorder[leftCount:l - 1])
        return TreeNode(val, left, right)

File: LeetCode/889. 根据前序和后序遍历构造二叉树.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn construct_from_pre_post(preorder: Vec<i32>, postorder: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
        match preorder.len() {
            0 => None,
            1 => Some(Rc::new(RefCell::new(TreeNode::new(preorder[0])))),
            l => {
                let val = preorder[0];
                let mut node = TreeNode::new(val);
                let val_of_left_child = preorder[1];
                let postorder_index = postorder.iter().position(|&x| x == val_of_left_child).unwrap();
                let left_count = postorder_index + 1;
                let left = Solution::construct_from_pre_post((&preorder[1..left_count + 1]).to_vec(), (&postorder[..left_count]).to_vec());
                let right = Solution::construct_from_pre_post((&preorder[left_count + 1..]).to_vec(), (&postorder[left_count..l - 1]).to_vec());
                node.left = left;
                node.right = right;
                Some(Rc::new(RefCell::new(node)))
            }
        }
    }
}
